src/aws-codeartifact/repository.ts

```diff
@@ -0,0 +1,248 @@
+      repositoryResourceArnParts.account === undefined ||
+      repositoryResourceArnParts.resourceName === undefined
+    ) {
+      throw new Error(`Expected a repository ARN, but got ${repositoryArn}`);
+    }
+    const repositoryNameParts = repositoryResourceArnParts.resourceName.split('/');
+    const domainName = repositoryNameParts[0];
+    const repositoryName = repositoryNameParts[1];
+
+    const domain = Domain.fromDomainArn(
+      scope,
+      'domain',
+      Stack.of(scope).formatArn({
+        resource: 'domain',
+        service: 'codeartifact',
+        resourceName: domainName,
+      }),
+    );
+
+    return Repository.fromRepositoryAttributes(scope, id, {
+      repositoryArn: repositoryArn,
+      repositoryName: repositoryName,
+      domain,
+    });
+  }
+
+  protected cfnResource: CfnRepository;
+  protected cfnResourceProps: CfnRepositoryProps;
+
+  readonly cdkTagManager: TagManager;
+  readonly repositoryArn: string;
+  readonly repositoryName: string;
+  readonly domain: IDomain;
+  /**
+   * Optional policy document that represents the resource policy of this repository
+   *
+   * If specified, addToResourcePolicy can be used to edit this policy.
+   * Otherwise this method will no-op.
+   */
+  protected readonly policy?: PolicyDocument;
+
+  constructor(scope: Construct, id: string, props: RepositoryProps) {
+    super(scope, id);
+    this.policy = new PolicyDocument();
+
+    this.cdkTagManager = new TagManager(TagType.KEY_VALUE, 'AWS::CodeArtifact::Repository');
+
+    this.cfnResourceProps = {
+      domainName: props.domain.domainName,
+      repositoryName: props.repositoryName,
+      description: props.description,
+      domainOwner: props.domain.domainOwner,
+      tags: this.cdkTagManager.renderTags(),
+      externalConnections: props.externalConnection !== undefined ? [props.externalConnection] : undefined, // only 1 allowed
+      permissionsPolicyDocument: this.policy,
+      upstreams: undefined, // TODO: add upstreams
+    };
+
+    this.cfnResource = this.createCfnResource();
+
+    this.repositoryArn = this.cfnResource.attrArn;
+    this.repositoryName = this.cfnResource.attrName;
+    this.domain = Domain.fromDomainArn(
+      this,
+      'domain',
+      Stack.of(this).formatArn({
+        resource: 'domain',
+        service: 'codeartifact',
+        account: this.cfnResource.attrDomainOwner,
+        resourceName: this.cfnResource.attrDomainName,
+      }),
+    );
+  }
+
+  protected createCfnResource(): CfnRepository {
+    return new CfnRepository(this, 'Resource', this.cfnResourceProps);
+  }
+
+  /**
+   * Adds a statement to the KMS key resource policy.
+   * @param statement The policy statement to add
+   * @param allowNoOp If this is set to `false` and there is no policy
+   * defined (i.e. external key), the operation will fail. Otherwise, it will
+   * no-op.
+   */
+  public addToResourcePolicy(statement: PolicyStatement, allowNoOp = true): AddToResourcePolicyResult {
```

Comment:
Why is `allowNoOp` necessary? If the user doesn't need it for that policy, wouldn't it not be called?

Reply:
I guess I did some copy-:spaghetti: and this is a probably still some left-over code from when this method was declared in the base class....

will delete it